### PR TITLE
[CRIMAP-580] Send supporting evidence data to datastore 

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change
This PR is part of a wider piece of work to serialise application evidence data. In this PR, `supporting_evidence` (aka document) data is serialised and sent to the datastore. The `filename`, `s3_object_key`, `file_size` `content_type` metadata is being serialised to start with. The PR also adds the functionality to rehydrate document data for resubmitted applications. 

## Link to relevant ticket
[CRIMAP-580](
https://dsdmoj.atlassian.net/jira/software/projects/CRIMAP/boards/897?selectedIssue=CRIMAP-580)

#### Submitting an application with evidence and screenshot of submitted application data showing the supporting_evidence attribute
https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/d829fbbd-7e10-4795-b28d-b2546f5058bc

Screenshot 2023-09-22 at 16 00 32

#### Resubmitting an application and checking the evidence data is repopulated
https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/bbe8ee1f-c8e9-4105-937b-1f695ceced63

## How to manually test the feature

Local setup 
In your local apply repo, checkout this branch and point to the updated version of the schema in the `Gemfile`
```
gem 'laa-criminal-legal-aid-schemas',
    github: 'ministryofjustice/laa-criminal-legal-aid-schemas', ref: 'cbc02594e7ed9ff33b6e20a371a054a3efce2d87'

```
Run `bundle install` and spin up apply with `rails server`

Then in your local datastore repo, checkout to [this branch](https://github.com/ministryofjustice/laa-criminal-applications-datastore/pull/170) and follow the same instructions above to change which version of the schema you are pointing to. Run `bundle install` and spin up datastore with `rails server`. In a second terminal, run localstack with `docker-compose up localstack` to allow you to upload documents. 

Navigate to review and simply spin up review with `rails server`. 

Now that you have set up your local environment, you can test by creating a draft application in apply and before submitting the application, navigating to `/steps/evidence/upload` to upload a document. Then review and submit the application. After submitting the application, open up a new terminal in your datastore repo and run `rails c`. Then get the id of the application in review and query the submitted application json with `CrimeApplication.find('2f4b3135-2203-4ba5-b497-5eb3b0632efc').submitted_application`. You should see document meta data. 

Next to test that resubmitted applications are populated with their previous document records in review, send that application back to the provider. Then in apply, navigate to the returned applications tab and click to update the application. Then find the resubmitted application and go to `/steps/evidence/upload` to see the document listed in the uploaded files section.

It is also worth testing an application submitted without any documents and submitting an application with a document that failed to upload correctly to ensure documents that errored are not included in the application json nor are they listed for a resubmitted application.

[CRIMAP-580]: https://dsdmoj.atlassian.net/browse/CRIMAP-580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/document_bundle.rb [hunk 1]
```diff
@@ -1,6 +1,10 @@
 class DocumentBundle < ApplicationRecord
   belongs_to :crime_application
-  has_many :documents, dependent: :destroy
+  has_many :documents, dependent: :destroy do
```
Comment (zheileman) [inline app/models/document_bundle.rb:3]: Maybe this is simpler?

```ruby
has_many :documents, dependent: :destroy
has_many :stored_documents, -> { where.not(s3_object_key: nil) }, class_name: 'Document'
```

We would need both, as there are use-cases where we want all documents to be returned, including those not persisted to S3 (for example in the upload page when reloading in non-JS version, to see errors).

It also avoids adding a method to the `Document` model, and might be easier to stub.

Comment (zheileman) [reply app/models/document_bundle.rb:3]: In addition to the above comment, there is another way, I think the simplest one, which is to add a scope to the `Document` model:

```ruby
scope :stored, -> { where.not(s3_object_key: nil) }
```

That way we can do:

```ruby
Document.stored

bundle = DocumentBundle.last
bundle.documents.stored
```

In any case at some point we will need a refactor to remove the concept of bundle as it seems will not longer be needed 😄

Comment (hiboabd) [reply app/models/document_bundle.rb]: That's much cleaner 🤩 I've made those changes

Comment (zheileman) [reply app/models/document_bundle.rb]: Nice! 🥳

**Review by zheileman — APPROVED** [review]
Great! 👍

**Pull request merged**